src/aspire/source/xform.py

```diff
@@ -0,0 +1,388 @@
+        """
+        super().__init__(resolution=resolution)
+        self.shifts = shifts
+        self.n = shifts.shape[0]
+
+    def _forward(self, im, indices):
+        return im.shift(self.shifts[indices])
+
+    def _adjoint(self, im, indices):
+        return im.shift(-self.shifts[indices])
+
+    def downsample(self, resolution):
+        self.shifts /= self.resolution / resolution
+        super().downsample(resolution)
+
+
+class DownSample(OneWayXform):
+    """
+    A Xform that downsamples a 3D Image to a resolution specified by this Xform's resolution.
+    """
+    def _forward(self, im, indices):
+        return im.downsample(self.resolution)
+
+
+class Filter(SymmetricXform):
+    """
+    A Xform that applies a single CTF Filter object to a 3D Image.
+    """
+    def __init__(self, filter, resolution=np.inf):
+        """
+        Initialize the Filter Xform using a Filter object
+        :param filter: An object of type `aspire.utils.filters.Filter`
+        :param resolution: Resolution of images expected to pass through this Transform
+        """
+        super().__init__(resolution=resolution)
+        self.filter = filter
+
+    def _forward(self, im, indices):
+        return im.filter(self.filter)
+
+    def downsample(self, resolution):
+        self.filter.scale(self.resolution / resolution)
+        super().downsample(resolution)
+
+
+class NoiseAdder(OneWayXform):
+    """
+    A Xform that adds white noise, optionally passed through a Filter object, to all incoming images.
+    """
+    def __init__(self, resolution=np.inf, seed=0, noise_filter=None):
+        """
+        Initialize the random state of this NoiseAdder using specified values.
+        :param resolution: Resolution of images expected to pass through this Transform
+        :param seed: The random seed used to generate white noise
+        :param noise_filter: An optional aspire.utils.filters.Filter object to use to filter the generated white noise.
+            Be default, a ScalarFilter is used, emulating true white noise, but any additional filter can be used
+            to emulate isotropic noise etc.
+        """
+        super().__init__(resolution=resolution)
+        self.seed = seed
+        self.noise_filter = noise_filter or ScalarFilter(value=1)
+
+    def _forward(self, im, indices):
+        im = im.copy()
+
+        for i, idx in enumerate(indices):
+            # Note: The following random seed behavior is directly taken from MATLAB Cov3D code.
+            random_seed = self.seed + 191 * (idx + 1)
+            im_s = randn(2 * self.resolution, 2 * self.resolution, seed=random_seed)
+            im_s = Image(im_s).filter(self.noise_filter, power=0.5)[:, :, 0]
+            im[:, :, i] += im_s[:self.resolution, :self.resolution]
+
+        return im
+
+
+class IndexedXform(Xform):
+    """
+    An IndexedXform is a Xform where individual Xform objects are used at specific indices of the incoming Image object.
```

Comment:
This might be violated at a bunch of places. Let's keep this under consideration in issue #69. In general I find 80 chars too restrictive. Closing here.